Makefile

```diff
@@ -1,9 +1,359 @@
+	@echo ' make clean && CC=gcc LD=g++ make all-c-tests'
+	@echo ' make run-c-tests'
+	@echo ' '
+	@echo 'To build C++-sample programs and run unit-tests:'
+	@echo ' make clean-l3 && CC=g++ CXX=g++ LD=g++ make all-cpp-tests'
+	@echo ' make run-cpp-tests'
+	@echo ' '
+	@echo 'To build C++-sample .cc programs and run unit-tests:'
+	@echo ' make clean-l3 && CC=g++ CXX=g++ LD=g++ make all-cc-tests'
+	@echo ' make run-cc-tests'
+	@echo ' '
+	@echo 'Environment variables: '
+	@echo ' BUILD_MODE={release,debug}'
+	@echo ' BUILD_VERBOSE={0,1}'
+
+#
+# Verbosity
+#
+ifndef BUILD_VERBOSE
+   BUILD_VERBOSE=0
+endif
+
+# Setup echo formatting for messages.
+ifeq "$(BUILD_VERBOSE)" "1"
+   COMMAND=
+   PROLIX=@echo
+   BRIEF=@ >/dev/null echo
+   # Always print message describe step executed, even in verbose mode.
+   # BRIEF_FORMATTED=@ >/dev/null echo
+   BRIEF_FORMATTED=@printf
+   BRIEF_PARTIAL=@echo -n >/dev/null
+else ifeq "$(BUILD_VERBOSE)" "0"
+   COMMAND=@
+   PROLIX=@ >/dev/null echo
+   BRIEF=@echo
+   BRIEF_FORMATTED=@printf
+   BRIEF_PARTIAL=@echo -n
+else
+   $(error Unknown BUILD_VERBOSE mode "$(BUILD_VERBOSE)".  Valid values are "0" or "1". Default is "0")
+endif
+
+# Compilers to use
+CC  ?= gcc
+CXX ?= g++
+LD  ?= gcc
+
+# ###################################################################
+# SOURCE DIRECTORIES AND FILES, Generator Package
+# ###################################################################
+#
+L3PACKAGE   = l3
+SRCDIR      = src
+INCDIR      = include
+L3_SRCDIR   = $(SRCDIR)
+L3_INCDIR   = $(INCDIR)
+TESTSDIR    = test-use-cases
+
+# ###################################################################
+# BUILD DIRECTORIES AND FILES
+# ###################################################################
+#
+ifndef BUILD_ROOT
+   BUILD_ROOT := build
+endif
+
+#
+# Build mode
+#
+ifndef BUILD_MODE
+   BUILD_MODE=release
+endif
+BUILD_DIR := $(BUILD_MODE)
+
+BUILD_PATH=$(BUILD_ROOT)/$(BUILD_DIR)
+
+OBJDIR = $(BUILD_PATH)/obj
+BINDIR = $(BUILD_PATH)/bin
+
+# ###################################################################
+# SOURCE DIRECTORIES AND FILES
+# ###################################################################
+
+L3_SRC      := $(L3_SRCDIR)/l3.c
+L3_ASSEMBLY := l3.S
+L3_SRCS     := $(L3_SRC) $(L3_ASSEMBLY)
+
+# ###################################################################
+# Symbols for test-case data files generated by sample programs
+# ###################################################################
+TMPDIR              := /tmp
+TEST_DATA_SUFFIX    := test.dat
+
+L3_C_DATA           := $(TMPDIR)/$(L3PACKAGE).c-$(TEST_DATA_SUFFIX)
+L3_C_SMALL_DATA     := $(TMPDIR)/$(L3PACKAGE).c-small-$(TEST_DATA_SUFFIX)
+L3_CPP_DATA         := $(TMPDIR)/$(L3PACKAGE).cpp-$(TEST_DATA_SUFFIX)
+L3_CPP_SMALL_DATA   := $(TMPDIR)/$(L3PACKAGE).cpp-small-$(TEST_DATA_SUFFIX)
+L3_CC_DATA          := $(TMPDIR)/$(L3PACKAGE).cc-$(TEST_DATA_SUFFIX)
+L3_CC_SMALL_DATA    := $(TMPDIR)/$(L3PACKAGE).cc-small-$(TEST_DATA_SUFFIX)
```

Comment:
This collection of test data names are also chosen with `l3` prefix, so `ls /tmp/l3*` will find them easily.

The sample programs are required to supply these names, which they do now.

All tested in CI.